music_rnn/src/model.ts

```diff
@@ -131,41 +155,160 @@ export class MusicRNN {
+    const length:number = inputs.shape[0];
+    const outputSize:number = inputs.shape[1];
+
+    let c: tf.Tensor2D[] = [];
+    let h: tf.Tensor2D[] = [];
+    for (let i = 0; i < this.biasShapes.length; i++) {
+      c.push(tf.zeros([1, this.biasShapes[i] / 4]));
+      h.push(tf.zeros([1, this.biasShapes[i] / 4]));
+    }
+
+    // Initialize with input.
+    const samples: tf.Tensor1D[] = [];
+    for (let i = 0; i < length + steps; i++) {
+      let nextInput: tf.Tensor2D;
+      if (i < length) {
+        nextInput = inputs.slice([i, 0], [1, outputSize]).as2D(1, outputSize);
+      } else {
+        const logits = h[h.length - 1].matMul(this.lstmFcW).add(this.lstmFcB);
+        const sampledOutput = (
+          temperature ?
+          tf.multinomial(tf.softmax(logits.div(tf.scalar(temperature))), 1)
+            .as1D():
+          logits.argMax(1).as1D());
+        nextInput = tf.oneHot(sampledOutput, outputSize).toFloat();
+        // Save samples as bool to reduce data sync time.
+          samples.push(nextInput.as1D().toBool());
+      }
+      [c, h] = tf.multiRNNCell(this.lstmCells, nextInput, c, h);
+    }
+    return samples;
+  }
+
+  private sampleRnnWithAttention(inputs: tf.Tensor2D,
+                                 steps: number,
+                                 temperature: number) {
+    const length:number = inputs.shape[0];
+    const outputSize:number = inputs.shape[1];
+
+    let c: tf.Tensor2D[] = [];
+    let h: tf.Tensor2D[] = [];
+    for (let i = 0; i < this.biasShapes.length; i++) {
+      c.push(tf.zeros([1, this.biasShapes[i] / 4]));
+      h.push(tf.zeros([1, this.biasShapes[i] / 4]));
+    }
+    const attnSize = this.biasShapes[0] / 4;
+    const attnLength = 32;
+    const numCounterBits = 6;
+
+    let attention: tf.Tensor1D = tf.zeros([attnSize]);
+    let attentionState: tf.Tensor2D = tf.zeros([1, attnSize * attnLength]);
+    let lastOutput: tf.Tensor2D;
+
+    // Initialize with input.
+    const samples: tf.Tensor1D[] = [];
+    for (let i = 0; i < length + steps; i++) {
+      let nextInput: tf.Tensor2D;
+      if (i < length) {
+        nextInput = inputs.slice([i, 0], [1, outputSize]).as2D(1, outputSize);
+      } else {
+        const logits = lastOutput.matMul(this.lstmFcW).add(this.lstmFcB);
+        const sampledOutput = (
+          temperature ?
+          tf.multinomial(logits.div(tf.scalar(temperature)), 1).as1D():
+          logits.argMax(1).as1D());
+        nextInput = tf.oneHot(sampledOutput, outputSize).toFloat();
+        // Save samples as bool to reduce data sync time.
+        samples.push(nextInput.as1D().toBool());
+      }
+
+      const binaryCounts = tf.buffer([1, numCounterBits]);
+      for (let bit = 0; bit < numCounterBits; bit++) {
+        const counterValue = Math.floor((i + 1) / 2 ** bit) % 2 ? 1.0 : -1.0;
+        binaryCounts.set(counterValue, 0, bit);
+      }
+      nextInput = nextInput.concat(binaryCounts.toTensor(), 1) as tf.Tensor2D;
```

Comment:
The binary counter should be in a data converter rather than here IMO, but need to address how to apply it when using the last output as the next input. We don't use the data converters for that.